User configurable latex lsp

PZ31k0nauT opened the pull request:
Added the possibility for user configuration for the latex-lsp(texlab).

The only hardcoded parts are the arrays (filetypes, latex.build.args and latex.forward_search.args).

I couldn't get it to work correctly otherwise.

I also added TexlabBuild and TexlabForward to which-key.

It should be mentioned that TexLab with the update to >3.0.0 has breaking changes. The setting variables are renamed and reorganized.
If there are problems the latex-lsp should be updated with `:LspInstall latex`.
Because of some conflicts I had to delete ~/.local/share/nvim/lspinstall/latex folder and make a fresh install.

Tastyep commented on ftplugin/tex.lua:
```diff
@@ -1,4 +1,42 @@
+      bibtexFormatter = O.lang.latex.bibtex_formatter,
+      build = {
+        args = { '-pdf', '-interaction=nonstopmode', '-synctex=1', '%f' },
+        -- args = O.lang.latex.build.args,
```
Either we make it configurable or not. It is misleading to have a O.lang.latex.build.args while changing it doesn't affect the configuration. I am in favor of keeping the configuration entry (for both compiler and viewer).

PZ31k0nauT replied:
I would love to have it configurable. But some how the arrays of the table in default-config.lua and lv-config.lua are not recognized. I'm new to lua maybe I'm making something wrong. The arrays which are affected are filetypes, texlab.build.args and texlab.forward_search.args.

For texlab.forward_search.args I try to implement a workaround. My plan is to write if statements which depend on the pdf-viewer, like zathura and so on.

Tastyep commented:
I took some time to test this and it works great (even latexindent which was a pain), however, I didn't manage to build.

> I also added TexlabBuild and TexlabForward to which-key.

It doesn't appear in the diff, is it wanted ?

barun108 commented:
> I took some time to test this and it works great (even latexindent which was a pain), however, I didn't manage to build.
> 
> > I also added TexlabBuild and TexlabForward to which-key.
> 
> It doesn't appear in the diff, is it wanted ?

Gonna try it

PZ31k0nauT commented:
> I took some time to test this and it works great (even latexindent which was a pain), however, I didn't manage to build.
> 
> > I also added TexlabBuild and TexlabForward to which-key.
> 
> It doesn't appear in the diff, is it wanted ?

I don't know if you have the same problem I had. I had to remove the latex-lsp completely and then reinstall it. Since TexLab 3.0.0 the settings are renamed. I would love to know how to check the version of the installed lsp.

Yeah the which-key part got lost as I fetched the master. I will update the PR.

Tastyep commented:
> I don't know if you have the same problem I had. I had to remove the latex-lsp completely and then reinstall it. Since TexLab 3.0.0 the settings are renamed. I would love to know how to check the version of the installed lsp.

I don't know if there is a command for that, but you can go check manually:
~/.local/share/nvim/lspinstall/latex
then run ./texlab --version
Personally, I already had v3.2.0

> Which settings did you use for the build. As I set them they should all be the default settings.
I didn't manage to build, I'll give it another go

Tastyep commented on lua/default-config.lua:
```diff
@@ -200,7 +200,36 @@ O = {
       filetypes = { "php", "phtml" },
     },
-    latex = {},
+    latex = {
```
Could you please review these options again ?
Check
https://github.com/neovim/nvim-lspconfig/blob/master/lua/lspconfig/texlab.lua
https://github.com/latex-lsp/texlab/blob/master/docs/options.md

I didn't check everything , but I know for sure some of the options don't exist (aux_directory, bibtex_formatter)
Also, please keep a consistent order, put the single line entries on top (The one of lspconfig is great, please follow it)

PZ31k0nauT replied:
Ok. I will follow the lspconfig with the only destinction that the single line entries on top. The settings auxDirectory and bibtexFormatter also exist in the lspconfig. Only in the table they are called O.lang.latex.aux_directory and O.lang.latex.bibtex_formatter.
I looked at other settings and thought this might be a kind of naming convention.

PZ31k0nauT commented:
**Update**
I updated the PR. Which-key settings are implemented again and examples settings in the lv-config.example.lua.

I tried to follow the formatting rules as best as possible. I also made sure that I used the correct settings from https://github.com/neovim/nvim-lspconfig/blob/master/lua/lspconfig/texlab.lua and https://github.com/latex-lsp/texlab/blob/master/docs/options.md.

The only hardcoded parts are the filetypes and the build arguments. I simply can't get it to work otherwise. My luafu is to week.

I also added all the preview arguments into the configuration. The user only has to reference the used executable as mentioned in https://github.com/latex-lsp/texlab/blob/master/docs/previewing.md.

But this setup has one drawback. The correct preview options have to be set in this repository. So it may be beneficial to enable the user to setup his own previewer and preview arguments.

I could only test it with zathura.

Tastyep commented on ftplugin/tex.lua:
```diff
@@ -1,4 +1,64 @@
+elseif O.lang.latex.forward_search.executable == "/Applications/Skim.app/Contents/SharedSupport/displayline" then
+  preview_settings = skim_args
+elseif O.lang.latex.forward_search.executable == "" then
+  preview_settings = {}
```
remove `elseif O.lang.latex.forward_search.executable == "" then`

PZ31k0nauT replied:
Done.

Tastyep commented on ftplugin/tex.lua:
```diff
@@ -1,4 +1,64 @@
+      auxDirectory = O.lang.latex.aux_directory,
+      bibtexFormatter = O.lang.latex.bibtex_formatter,
+      build = {
+        args = { '-pdf', '-interaction=nonstopmode', '-synctex=1', '%f' },
```
I tested passing the configuration entry and can guarantee you that it does work.
Please update it

PZ31k0nauT replied:
Ok. You are right. For a moment I thought it works because it falls back to the defaults. But no.
This set me off:
![build_args_are_empty](https://user-images.githubusercontent.com/86895284/124570420-26353e80-de47-11eb-8cea-4562344a4b92.png)

I glad that it works with the configuration through lv-config.lua
And done. :)

Tastyep replied:
Yep I know, that was misleading for me too :)

Tastyep commented on utils/installer/lv-config.example.lua:
```diff
@@ -66,6 +66,24 @@ O.lang.php.diagnostics.signs = true
+--LaTeX
+O.lang.latex.aux_directory = "."
+O.lang.latex.bibtex_formatter = "texlab"
+--TODO build args are hardcoded; args are not passed through to ftplugin/tex.lua
```
To update

PZ31k0nauT replied:
Sorry. I forgot to update the build args.

Tastyep commented on utils/installer/lv-config.example.lua:
```diff
@@ -66,6 +66,24 @@ O.lang.php.diagnostics.signs = true
 O.lang.php.diagnostics.underline = true
 O.lang.php.filetypes = {"php", "phtml"}
 
+--LaTeX
```
Could you add this line below the comment
-- Reference: https://github.com/latex-lsp/texlab/blob/master/docs/options.md

PZ31k0nauT replied:
Done

Tastyep commented on utils/installer/lv-config.example.lua:
```diff
@@ -66,6 +66,24 @@ O.lang.php.diagnostics.signs = true
+-- see https://github.com/latex-lsp/texlab/blob/master/docs/previewing.md for options
+-- insert the executable as referenced in previewng.md; lunarvim will make the rest
+O.lang.latex.forward_search.executable = "zathura"
+O.lang.latex.latex_formatter = "latexindent"
```
This might be redundant with neoformat, but let's leave it for now and see how it goes

Tastyep reviewed (changes requested):
Looking good! Some small corrections and I'll approve it :)

PZ31k0nauT reviewed (commented):
I think that should do it. To be honest I don't know how I got into this review mode.

This whole PR thing is completely new to me. I hope I didn't make it to complicated.

PZ31k0nauT commented:
Just an update.

I resolved some merge conflicts and deleted the which-key entries. I think these can be inserted by the end user in lv-config.lua.

On question though. Does it make sense to keep this PR alive? I don't know it it's wanted or not.

I know a lot is going on, which is a pretty cool thing. So no pressure. Just wanted to know.

PZ31k0nauT commented:
Yeah I will close this pull request and open a new one. Somehow I got all those commit in here. There is a lot to learn in regards of the PR game.
